tests/test_encoding/test_fix_index_mismatch_from_upstream_array.py

```diff
@@ -0,0 +1,54 @@
+import numpy as np
+import pandas as pd
+import pytest
+from sklearn.impute import SimpleImputer
+
+from feature_engine.encoding import MeanEncoder, PRatioEncoder, WoEEncoder
+
+
+@pytest.mark.parametrize(
+    # Encoders that encode X as a function of y; this is what
+    # breaks down when X becomes an array and indexes don't accidentally match in final
+    # concantenation
+    "encoder",
+    [MeanEncoder(), WoEEncoder(), PRatioEncoder()],
+)
+def test_fix_index_mismatch_from_upstream_array(encoder):
+    """
+    Created 2022-02-19 to test fix to issue # 376
+    Code adapted from:
+    https://github.com/scikit-learn-contrib/category_encoders/issues/280
+    """
+
+    # test dataframe; setup for a transfromation where
+    # coded version of 'x' will be a function of target 'y'
+    df: pd.DataFrame = pd.DataFrame(
+        {
+            "x": ["a", "a", "b", "b", "c", "c"],
+            "y": [1, 0, 1, 0, 1, 0],
+        }
+    )
+    # Key - "non-standard" index that is not the usual
+    # contiguous range starting a t 0
+    df.index = [101, 105, 42, 76, 88, 92]
+
+    # Set up for standard pipeline/training etc.
+    X: pd.DataFrame = df[["x"]]
+    y: pd.Series = df["y"]
+
+    # Will serve as a no-op whose chief purpose is to turn the
+    # X into an np.ndarray
+    si = SimpleImputer(strategy="constant", fill_value="a")
```

Comment:
Can we not just convert the dataframe into an array without calling the simple imputer? it should be faster. Because here we are using the simple imputer just to reset the index and obtain an array from a dataframe. So we could in theory do this with numpy. And then we have less dependencies.

In fact, we could do the following instead:

-  create a dataframe X with some index and a y is an array
- create an array X and y is a series.

And we test both, avoiding the simple imputer and the assert types below. Am I correct?

Reply:
Definitely agreed about getting rid of SimpleImputer and shortening code; done in latest commit.

However, in terms of this case you mentioned:
* create a dataframe X with some index and a y is an array
Under what circumstances could y get changed from a Series into an array into the course of sklearn transforms? As far as I can tell no transforms affect y, though I could be mistaken?

Reply:
I am thinking, if a user created a target like this `y = np.where(df['my_var']=='yes', 1, 0)`, that would return an array. 

The truth is, I am not sure how often this happens. But at the moment, most of our transformers do not enforce y to be a series. So users could, in practice, pass a numpy array. So it might be worth to expand the check, so that transformers work for them as well.

But then again, I am not so certain about how often this could happen. What is your experience @bmreiniger @noahjgreen295 ?

Reply:
Alright, I did some research: most of our transformers that take y, would use a Scikit-learn estimator at the back, so for those, the `Check_X_y` from sklearn would take care of X and y and we have nothing to worry about.

The only transformers that might have a problem are the encoders, woe, meantargetencoder, probabilityratio, like @noahjgreen295 sayd and I think the ordinal(encoding_method="ordered") should have it too (more below). 

Those transformers, if they receive an array for y, they would transform it to a series, and then it would be reindexed from 0. Likely, the index of the converted series will not match the train df index. But if this is the case, the logic in the check_x_y_mismatch as it is now, should handle it. But, note that we would be giving the df an index that is reindexed from 0.

So I would say, let's add a test for when X is dataframe, and y is an array, just to corroborate that what I am saying is true. But I don't think we need to change the logic of the check_x_y_mismatch.

Also, the ordinalencoder(encoding_method="ordered") uses exactly the same logic as all other encoders. So I am not sure why that transformer would not fail the test if we do not add the check_x_y mistmatch. If you added a test to see that it passes, that would be great.